src/main/java/blusunrize/immersiveengineering/api/crafting/FluidTagInput.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * BluSunrize
+ * Copyright (c) 2020
+ *
+ * This code is licensed under "Blu's License of Common Sense"
+ * Details can be found in the license file in the root folder of this project
+ *
+ */
+
+package blusunrize.immersiveengineering.api.crafting;
+
+import com.google.common.base.Preconditions;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import net.minecraft.fluid.Fluid;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.nbt.JsonToNBT;
+import net.minecraft.network.PacketBuffer;
+import net.minecraft.tags.FluidTags;
+import net.minecraft.tags.Tag;
+import net.minecraft.util.JSONUtils;
+import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.common.util.Lazy;
+import net.minecraftforge.fluids.FluidStack;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+public class FluidTagInput implements Predicate<FluidStack>
+{
+	protected final Lazy<Tag<Fluid>> fluidTag;
+	protected final int amount;
+	protected final CompoundNBT nbtTag;
+
+	public FluidTagInput(Lazy<Tag<Fluid>> fluidTag, int amount, CompoundNBT nbtTag)
+	{
+		this.fluidTag = fluidTag;
+		this.amount = amount;
+		this.nbtTag = nbtTag;
+	}
+
+	public FluidTagInput(ResourceLocation resourceLocation, int amount, CompoundNBT nbtTag)
+	{
+		this(Lazy.of(() -> new FluidTags.Wrapper(resourceLocation)), amount, nbtTag);
+	}
+
+	public FluidTagInput(ResourceLocation resourceLocation, int amount)
+	{
+		this(resourceLocation, amount, null);
+	}
+
+	public static FluidTagInput deserialize(JsonElement input)
+	{
+		Preconditions.checkArgument(input instanceof JsonObject, "FluidTagWithSize can only be deserialized from a JsonObject");
+		JsonObject jsonObject = input.getAsJsonObject();
+		ResourceLocation resourceLocation = new ResourceLocation(JSONUtils.getString(jsonObject, "tag"));
+		if(!JSONUtils.hasField(jsonObject, "nbt"))
+			return new FluidTagInput(resourceLocation, JSONUtils.getInt(jsonObject, "amount"));
+		try
+		{
+			CompoundNBT nbt = JsonToNBT.getTagFromJson(JSONUtils.getString(jsonObject, "nbt"));
```

Comment:
Somethinglike this (Pulled from CraftingHelper) would allow for users doing packs to not have to write an escape NBT string (Yes I know for Datagens it's not awful)
```                
if(element.isJsonObject())
  nbt = JsonToNBT.getTagFromJson(GSON.toJson(element));
else
  nbt = JsonToNBT.getTagFromJson(JSONUtils.getString(element, "nbt"));
```

Reply:
Fair point. I was matching vanilla handling here.